### PR TITLE
security_layer: support ssl_context_factory argument on make

### LINKED ISSUE #283: make_security_layer should allow passing an ssl context factory
The tls_with_password_based_authentication had it, and it’s needed in some cases such as #282.

### PR DESCRIPTION
**Pull request by horazont**
Fixes #283.

### CODE REVIEW

File: aioxmpp/security_layer.py [hunk 1]
```diff
@@ -1376,6 +1381,12 @@ def make(
     :data:`None` while `pin_store` is not, a coroutine which returns
     :data:`False` is substituted.
 
+    `ssl_context_factory` can be a callable taking no arguments and returning
+    a :class:`OpenSSL.SSL.Context` object. If given, the factory will be used
+    to obtain an SSL context when the stream negotiates transport layer security
+    via TLS. By default, :func:`aioxmpp.security_layer.default_ssl_context` is
+    used, which should be fine for most applications.
```
Comment (sebastianriese) [inline aioxmpp/security_layer.py]: This should sound more like the hdparm manual page; we should note that you should only use another context if you know what you are doing as using a misconfigured context can disable the authentication of the peer.

File: tests/test_security_layer.py [hunk 2]
```diff
@@ -2425,6 +2425,52 @@ def test_simple(self):
...
+                )
+            )
+
+            PKIXCertificateVerifier = stack.enter_context(
+                unittest.mock.patch(
+                    "aioxmpp.security_layer.PKIXCertificateVerifier"
+                )
+            )
+
+            default_ssl_context = stack.enter_context(
```
[the remark this answers is not in the document]
Comment (horazont) [reply tests/test_security_layer.py]: I thought that too (I copied the code from the other test case), but then I realized that there is in fact a semantic difference which can only be caught by mocking the function: if one sets the ``ssl_context_factory=default_ssl_factory`` at function definition time in the parameter list, an application cannot monkey patch ``default_ssl_factory``.

But you’re right that it isn’t even needed in this test case, so it should be dropped.

File: aioxmpp/security_layer.py [hunk 3]
```diff
@@ -1398,22 +1428,36 @@ def make(
...
-       Take the security and privacy considerations from :rfc:`4505` (which
-       specifies the ANONYMOUS SASL mechanism) and :xep:`175` (which discusses
-       the ANONYMOUS SASL mechanism in the XMPP context) into account before
-       using `anonymous`.
+        Take the security and privacy considerations from :rfc:`4505` (which
+        specifies the ANONYMOUS SASL mechanism) and :xep:`175` (which discusses
+        the ANONYMOUS SASL mechanism in the XMPP context) into account before
+        using `anonymous`.
 
     The versaility and simplicity of use of this function make (pun intended)
```
Comment (sebastianriese) [inline aioxmpp/security_layer.py]: If you are cleaning up the docs anyways: "vesaility" => "versatility"

Comment (horazont) [reply aioxmpp/security_layer.py]: Done!

**Pull request merged**